File: core/src/ledger.rs
//! The decision ledger: safe reward handling — the port of `ledger.py`.
//!
//! Every open decision resolves in exactly one of three ways — rewarded,
//! expired(default_reward) at the horizon, or censored — each a deliberate,
//! loggable event. Resolving removes the record, so duplicates are
//! structural no-ops; no code path learns from an open decision. Resolution
//! order within one `expire` sweep is ledger (insertion) order.

use crate::decide::{BanditState, DecisionRecord};
use crate::model::update;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Kind {
    Rewarded,
    Expired,
    Censored,
}

impl Kind {
    pub fn as_str(self) -> &'static str {
        match self {
            Kind::Rewarded => "rewarded",
            Kind::Expired => "expired",
            Kind::Censored => "censored",
        }
    }
}

/// One deliberate, loggable resolution event. `reward` is the value the
/// model trained with; `None` for censored (excluded from training).
#[derive(Clone, Debug, PartialEq)]
pub struct Resolution {
    pub decision_id: String,
    pub kind: Kind,
    pub reward: Option<f64>,
}

/// The open record with this ID, removed from the ledger; `None` (ledger
/// untouched) if it is unknown or already resolved.
fn take(ledger: &mut Vec<DecisionRecord>, decision_id: &str) -> Option<DecisionRecord> {
    let i = ledger.iter().position(|r| r.decision_id == decision_id)?;
    Some(ledger.remove(i))
}

/// Resolve an open decision as rewarded(reward). Unknown or already
/// resolved IDs are a no-op returning `None`.
pub fn learn(state: &mut BanditState, decision_id: &str, reward: f64) -> Option<Resolution> {
    let record = take(&mut state.ledger, decision_id)?;
    update(&mut state.model, &record.features, reward);
    state.model_version += 1;
    Some(Resolution {
        decision_id: record.decision_id,
        kind: Kind::Rewarded,
        reward: Some(reward),
    })
}

/// Resolve an open decision as censored: removed from the ledger without
/// training, the exclusion itself returned for the log.
pub fn censor(state: &mut BanditState, decision_id: &str) -> Option<Resolution> {
    let record = take(&mut state.ledger, decision_id)?;
    Some(Resolution {
        decision_id: record.decision_id,
        kind: Kind::Censored,
        reward: None,
    })
}

/// Resolve every open decision whose horizon has passed at time `t` as
/// expired(default_reward), in ledger order. Callers sweep this with each
/// event's time; that sweep is what keeps the ledger bounded.
pub fn expire(state: &mut BanditState, t: f64) -> Vec<Resolution> {
    let mut resolutions = Vec::new();
    let mut remaining = Vec::with_capacity(state.ledger.len());
    for record in std::mem::take(&mut state.ledger) {
        if record.t + state.horizon <= t {
            update(&mut state.model, &record.features, state.default_reward);
            resolutions.push(Resolution {
                decision_id: record.decision_id,
                kind: Kind::Expired,
                reward: Some(state.default_reward),
            });
        } else {
            remaining.push(record);
        }
    }
    state.ledger = remaining;
    state.model_version += resolutions.len() as u64;
    resolutions
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::decide::{decide, new_bandit};
    use crate::exploration::DEFAULT_EPSILON;
    use crate::model::DEFAULT_FORGETTING;

    /// The golden episode resolves each decision exactly once (its
    /// `rejected` attempts cover the no-op paths through the final-state
    /// comparison); this pins the same exactly-once property directly, the
    /// way the reference's pytest suite does: every second resolution —
    /// duplicate, conflicting, post-expiry, post-censor, unknown — returns
    /// None and leaves every bit of the state alone.
    #[test]
    fn resolutions_are_exactly_once() {
        let mut state = new_bandit(4, 10.0, 0.0, DEFAULT_EPSILON, DEFAULT_FORGETTING).unwrap();
        let candidates = vec![vec![(0, 1.0)], vec![(1, 1.0)]];
        let a = decide(&mut state, &candidates, 0.0, "test", None, None).unwrap();
        let b = decide(&mut state, &candidates, 1.0, "test", None, None).unwrap();
        let c = decide(&mut state, &candidates, 2.0, "test", None, None).unwrap();

        assert!(learn(&mut state, &a.decision_id, 1.0).is_some());
        assert!(censor(&mut state, &b.decision_id).is_some());
        let expired = expire(&mut state, 12.0); // c is due at exactly 12.0
        assert!(expired.len() == 1 && expired[0].decision_id == c.decision_id);
        assert!(state.ledger.is_empty() && state.model_version == 2);

        let before = state.clone();
        assert!(learn(&mut state, &a.decision_id, 0.0).is_none()); // conflicting duplicate
        assert!(learn(&mut state, &b.decision_id, 1.0).is_none()); // reward after censor
        assert!(learn(&mut state, &c.decision_id, 1.0).is_none()); // reward after expiry
        assert!(censor(&mut state, &a.decision_id).is_none()); // censor after reward
        assert!(learn(&mut state, "never-made:0", 1.0).is_none()); // unknown id
        assert!(expire(&mut state, 100.0).is_empty()); // nothing due
        assert!(state == before, "a rejected resolution moved the state");
    }
}

File: src/gittins_reference/ledger.py
"""The decision ledger: safe reward handling (design doc section 6).

The engine owns the join between decisions and rewards. The state's ledger
holds every *open* decision record — made by `decide`, not yet resolved —
and the only way to train the model is to resolve one of them. Every open
decision resolves in exactly one of two ways, each a deliberate event
returned to the caller for logging:

    learn(state, decision_id, reward, t)  ->  rewarded(reward), or
                                              expired(default_reward) when t
                                              says the horizon already passed
    expire(state, t)                      ->  expired(default_reward), for every
                                              decision past its horizon

Three properties, by construction rather than by discipline:

- **Idempotent** — resolving removes the record from the ledger, so a
  duplicate (or conflicting, or bogus) report finds nothing and is a no-op.
  There is one record per decision and it can be spent once.
- **Late rewards are safe, in arrival order** — rewards may arrive late,
  out of order, or never, and every case resolves through the same two
  paths with nothing lost or double-counted. Training applies at the
  *resolution's* position in the update sequence with the model's current
  forgetting weight (model.py has no notion of time), so a late reward
  counts slightly less than an on-time one would have — the model is
  simply trained in the order the world reported outcomes. Replaying a
  model therefore requires the ordered resolution sequence, which the
  decision log provides.
- **The horizon is enforced by the engine, not the sweep** — `learn`
  receives the caller's time and checks it against the open record's own
  decision time: a reward arriving at or past `record.t + horizon` is not
  a reward at all but an expiry that hadn't been swept yet, and it trains
  as `default_reward` exactly as `expire` would have. The declared cutoff
  therefore holds even if the caller's sweep cadence lags.
- **The classic silent bug is unrepresentable** — "reward hasn't arrived
  yet" cannot be read as "reward was zero", because an open decision is not
  training data and no code path learns from one. Absence becomes a zero (or
  whatever the application declared) only through `expire`, an explicit
  event at a declared horizon.

The ledger is bounded by the horizon: a decision is due `horizon` seconds
after it was made, so under a regular `expire` sweep (the caller's event
loop passes each event's time) the ledger never holds more than one
horizon's worth of open decisions. A reward that arrives after its decision
expired finds the ledger empty and is ignored — the horizon is the
application's declared cutoff, not a tunable.

Resolution order within one `expire` sweep is ledger (insertion) order,
fixed, so a sweep is bit-reproducible like everything else.
"""

from dataclasses import dataclass, replace

from gittins_reference.decide import BanditState, DecisionRecord
from gittins_reference.model import update

REWARDED = "rewarded"
EXPIRED = "expired"


@dataclass(frozen=True)
class Resolution:
    """One deliberate, loggable resolution event. `reward` is the value the
    model trained with."""

    decision_id: str
    kind: str  # REWARDED | EXPIRED
    reward: float


def take(
    ledger: "tuple[DecisionRecord, ...]", decision_id: str
) -> "tuple[DecisionRecord | None, tuple[DecisionRecord, ...]]":
    """The open record with this ID (or None) and the ledger without it."""
    for i in range(len(ledger)):
        if ledger[i].decision_id == decision_id:
            return ledger[i], ledger[:i] + ledger[i + 1 :]
    return None, ledger


def learn(
    state: BanditState, decision_id: str, reward: float, t: float
) -> "tuple[Resolution | None, BanditState]":
    """Resolve an open decision at time `t`: rewarded(reward) inside the
    horizon, expired(default_reward) at or past it — the same boundary
    `expire` uses, so a late reward can never train as a timely one.
    Unknown or already resolved IDs are a no-op: (None, unchanged state)."""
    record, rest = take(state.ledger, decision_id)
    if record is None:
        return None, state
    if record.t + state.horizon <= t:
        kind, trained = EXPIRED, state.default_reward
    else:
        kind, trained = REWARDED, reward
    new_state = replace(
        state,
        model=update(state.model, record.features, trained),
        model_version=state.model_version + 1,
        ledger=rest,
    )
    return Resolution(decision_id, kind, trained), new_state


def expire(state: BanditState, t: float) -> "tuple[tuple[Resolution, ...], BanditState]":
    """Resolve every open decision whose horizon has passed at time `t` as
    expired(default_reward), in ledger order. Callers sweep this with each
    event's time; that sweep is what keeps the ledger bounded."""
    resolutions = []
    remaining = []
    model = state.model
    for record in state.ledger:
        if record.t + state.horizon <= t:
            model = update(model, record.features, state.default_reward)
            resolutions.append(Resolution(record.decision_id, EXPIRED, state.default_reward))
        else:
            remaining.append(record)
    if not resolutions:
        return (), state
    new_state = replace(
        state,
        model=model,
        model_version=state.model_version + len(resolutions),
        ledger=tuple(remaining),
    )
    return tuple(resolutions), new_state
